### PR TITLE
Issue #624 Load rule parameters from additional files

### LINKED ISSUE #624: Load rule parameters from additional files

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #624

### CODE REVIEW

File: build/Version.props [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <MainVersion>1.6.0</MainVersion>
+    <MainVersion>1.5.1</MainVersion>
```
Comment (dbolkensteyn) [inline build/Version.props]: @tvsonar I don't think we're going to need a version change as part of this PR.

File: src/SonarLint.Vsix/Diagnostic.nuspec [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
   <metadata>
     <id>SonarLint</id>
-    <version>1.6.0</version>
+    <version>1.5.1</version>
```
Comment (dbolkensteyn) [inline src/SonarLint.Vsix/Diagnostic.nuspec]: @tvsonar same here, not required

File: src/SonarLint/Common/AnalyzerAdditionalFile.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        public AnalyzerAdditionalFile(string path)
+        {
+            this.path = path;
+        }
+
+        public override string Path => path;
+
+        public override SourceText GetText(CancellationToken cancellationToken)
+        {
+            return SourceText.From(File.OpenRead(path));
```
Comment (dbolkensteyn) [inline src/SonarLint/Common/AnalyzerAdditionalFile.cs]: @tvsonar in the doc of the `SourceText.From()` function, it says:

```
//     Reads from the beginning of the stream. Leaves the stream open.
```

Shouldn't we close & dispose of it then?

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Reflection;
+using System.IO;
+using System.Security;
+
+namespace SonarLint.Helpers
+{
+    public static class ParameterLoader
+    {
+        public static readonly string CheckedFileName = "sonarlint.xml";
```
Comment (dbolkensteyn) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @tvsonar why not SonarLint.xml ?

Comment (tvsonar) [inline src/SonarLint/Helpers/ParameterLoader.cs]: it doesn't really matter, we're using `StringComparison.InvariantCultureIgnoreCase` for the comparison.

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        private readonly static HashSet<DiagnosticAnalyzer> ProcessedAnalyzers = new HashSet<DiagnosticAnalyzer>();
+        private readonly static object lockObject = new object();
+
+        public static void SetParameterValues(DiagnosticAnalyzer parameteredAnalyzer,
+            AnalyzerOptions options)
+        {
+            var additionalFile = options.AdditionalFiles
+                .FirstOrDefault(f => f.Path.EndsWith(CheckedFileName, StringComparison.InvariantCultureIgnoreCase));
```
Comment (tvsonar) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @dbolkensteyn should we check exact file match here? we are just checking the path ending with sonarlint.xml.

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+            if (additionalFile == null || ProcessedAnalyzers.Contains(parameteredAnalyzer))
```
Comment (dbolkensteyn) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @tvsonar This isn't thread-safe.

HashSet is not thread-safe: "Any instance members are not guaranteed to be thread safe" https://msdn.microsoft.com/en-us/library/bb359438(v=vs.110).aspx

This means that calling instance methods from several threads concurrently might blow up the internal state of the object. For example, calling `Contains()` from multiple threads concurrently might not work. Whether or not it will actually work only depends on how `Contains()` is implemented - but the doc says you cannot assume this to be safe => It's not safe.

Now, a simple fix would be to replace it by a concurrent collection, whose instance methods are thread-safe, such as `ConcurrentDictionary` as suggested in http://stackoverflow.com/questions/18922985/concurrent-hashsett-in-net-framework

I think that would be perfectly acceptable solution, however it wouldn't be sufficient to guarantee that parameters for a given analyzer are loaded at most once. Indeed:

1. 2 threads can call `SetParameterValues()` on the same analyzer instance
2. They both call `Contains()` which returns false
3. They both proceed with the loading of their parameters
4. They acquire the lock to write the result back, the last one to execute won't update the set as the analyzer is already present

So you'd actually have to make the whole reading of the parameters part of the critical section if you'd wanted to ensure that it gets executed at most once.

To sum up, multi-threading is super hard to get right, especially when we implement it by ourselves by introducing locks and explicitly acquiring them and so on.

My :+1: goes for removing the lock, and replacing the `HashSet` by a `ConcurrentDictionary`, and not care that the same parameters could potentially be set several times on the same analyzer.

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+            }
+            catch (Exception exception)
+            {
+                if (exception is ArgumentNullException ||
+                    exception is SecurityException ||
+                    exception is FileNotFoundException ||
+                    exception is UriFormatException)
+                {
+                    // we expect these exceptions from the XML parsing.
```
Comment (dbolkensteyn) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @tvsonar how can any of these exception be expected? If the additional file property is set, then the additional file should exist.

There is already some code above to verify that there is a SonarLint XML additional file, and quit the function if that's not the case.

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                    return;
+                }
+
+                throw;
+            }
+            finally
+            {
+                CacheProcessedAnalyzer(parameteredAnalyzer);
```
Comment (dbolkensteyn) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @tvsonar It looks strange to cache from the finally

File: src/SonarQube.SonarLint.Descriptor/Descriptor.cs.nuspec [hunk 9]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
   <metadata>
     <id>SonarLint.Descriptor.CSharp</id>
-    <version>1.6.0</version>
+    <version>1.5.1</version>
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Descriptor/Descriptor.cs.nuspec]: as above, unnecessary

File: src/SonarQube.SonarLint.Descriptor/Descriptor.vbnet.nuspec [hunk 10]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
   <metadata>
     <id>SonarLint.Descriptor.VisualBasic</id>
-    <version>1.6.0</version>
+    <version>1.5.1</version>
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Descriptor/Descriptor.vbnet.nuspec]: as above, unnecessary

File: src/SonarQube.SonarLint.Runner/DiagnosticsRunner.cs [hunk 11]
```diff
@@ -57,7 +60,10 @@ public IEnumerable<Diagnostic> GetDiagnostics(Compilation compilation)
 
             using (var tokenSource = new CancellationTokenSource())
             {
-                var compilationWithAnalyzer = modifiedCompilation.WithAnalyzers(diagnosticAnalyzers, cancellationToken: tokenSource.Token);
+                var compilationWithAnalyzer = modifiedCompilation.WithAnalyzers(
+                    diagnosticAnalyzers,
+                    new AnalyzerOptions(ImmutableArray.Create<AdditionalText>(new AnalyzerAdditionalFile(configuration.Path))),
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Runner/DiagnosticsRunner.cs]: nice :)

File: src/SonarQube.SonarLint.Runner/Program.cs [hunk 12]
```diff
@@ -34,12 +33,27 @@ public static class Program
     {
         public static int Main(string[] args)
         {
+            if (args.Length != 3)
+            {
+                Write("Expected parameters: ");
+                Write("[Input configuration path: should end with sonarlint.xml, otherwise rule parameter value will be ignored]");
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Runner/Program.cs]: @tvsonar the ": should end with sonarlint.xml, otherwise rule parameter value will be ignored]" part can be removed - the fail fast behavior will make it explicit enough.

File: src/SonarQube.SonarLint.Runner/Program.cs [hunk 13]
```diff
@@ -34,12 +33,27 @@ public static class Program
...
+                Write("[Output file path]");
+                Write("[AnalyzerLanguage: 'cs' for C#, 'vbnet' for VB.Net]");
+
+                return -1;
+            }
+
+            if (!args[0].EndsWith(ParameterLoader.CheckedFileName, StringComparison.InvariantCultureIgnoreCase))
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Runner/Program.cs]: @tvsonar should be moved to `Configuration`, it's not the responsibility of the rule runner executable to know about this

File: src/SonarQube.SonarLint.Runner/Configuration.cs [hunk 14]
```diff
@@ -29,66 +29,39 @@
...
+        public string Path { get; private set; }
 
         public bool IgnoreHeaderComments { get; }
         public IImmutableList<string> Files { get; }
         public IImmutableSet<string> AnalyzerIds { get; }
 
-        public Configuration(XContainer xml, AnalyzerLanguage language)
+        public Configuration(string path, AnalyzerLanguage language)
         {
+            Path = path;
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Runner/Configuration.cs]: @tvsonar Here we should fail fast is the provided XML filename doesn't match `SonarLint.xml`

File: src/SonarQube.SonarLint.Runner/Runner.nuspec [hunk 15]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
   <metadata>
     <id>SonarLint.Runner</id>
-    <version>1.6.0</version>
+    <version>1.5.1</version>
```
Comment (dbolkensteyn) [inline src/SonarQube.SonarLint.Runner/Runner.nuspec]: not necessary

File: src/Tests/SonarLint.SonarQube.Integration.UnitTest/SonarLint.SonarQube.Integration.UnitTest.csproj [hunk 16]
```diff
@@ -132,7 +132,7 @@
   </ItemGroup>
   <ItemGroup>
     <None Include="app.config" />
-    <None Include="TestResources\ConfigurationTest.xml">
+    <None Include="TestResources\SonarLint.xml">
```
Comment (dbolkensteyn) [inline src/Tests/SonarLint.SonarQube.Integration.UnitTest/SonarLint.SonarQube.Integration.UnitTest.csproj]: @tvsonar here SonarLint.xml is already properly cased :)

File: src/Tests/SonarLint.UnitTest/Verifier.cs [hunk 17]
```diff
@@ -191,7 +192,10 @@ private static Document GetDocument(string filePath, string assemblyName, AdhocW
 
                 var compilationWithOptions = compilation.WithOptions(compilationOptions);
                 var compilationWithAnalyzer = compilationWithOptions
-                    .WithAnalyzers(ImmutableArray.Create(diagnosticAnalyzer), cancellationToken: tokenSource.Token);
+                    .WithAnalyzers(
+                        ImmutableArray.Create(diagnosticAnalyzer),
+                        new AnalyzerOptions(ImmutableArray.Create<AdditionalText>()),
```
Comment (dbolkensteyn) [inline src/Tests/SonarLint.UnitTest/Verifier.cs]: @tvsonar doesnt this have the exact same effect as before? what's the point of having more code to do the same thing?

File: src/SonarLint/Helpers/ParameterLoader.cs (hunk 8 above)
Comment (tvsonar) [inline src/SonarLint/Helpers/ParameterLoader.cs]: @dbolkensteyn we could do the caching even before we calculated anything. The thing is, we don't care if the parsing failed or succeeded, we don't want to do it again.

**Comment by dbolkensteyn** [comment]
See inline comments

File: src/SonarLint.VisualBasic/Rules/OptionalParameter.cs [hunk 18]
```diff
@@ -47,7 +47,9 @@ public class OptionalParameter : OptionalParameterBase<SyntaxKind, MethodBaseSyn
             parameter.Modifiers.First(m => m.IsKind(SyntaxKind.OptionalKeyword)).GetLocation();
 
         protected override IEnumerable<ParameterSyntax> GetParameters(MethodBaseSyntax method) =>
-            method.ParameterList.Parameters;
+            method.ParameterList == null
+            ? Enumerable.Empty<ParameterSyntax>()
+            : method.ParameterList?.Parameters;
```
Comment (dbolkensteyn) [inline src/SonarLint.VisualBasic/Rules/OptionalParameter.cs]: `?.` doesn't seem to be wanted here, as `ParameterList` is already checked for not being null. Otherwise it seems like `GetParameters` is expected under some condition to return null.

File: src/SonarLint/Helpers/ParameterLoader.cs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+            return builder.ToImmutable();
+        }
+
+        private readonly static ConcurrentDictionary<DiagnosticAnalyzer, byte> ProcessedAnalyzers = new ConcurrentDictionary<DiagnosticAnalyzer, byte>();
+
+        public static void SetParameterValues(DiagnosticAnalyzer parameteredAnalyzer,
+            AnalyzerOptions options)
+        {
+            var additionalFile = options.AdditionalFiles
```
Comment (dbolkensteyn) [inline src/SonarLint/Helpers/ParameterLoader.cs]: It could make sense to not even run this code if the analyzer is already cached.

**Comment by dbolkensteyn** [comment]
LGTM
